Add electrum_client in electrum_blockchain

danielabrozzoni opened the pull request.

danielabrozzoni committed:
Add electrum_client in electrum_blockchain

danielabrozzoni committed:
Add get_random_utxo, with_capacity to ElectrumClient

with_capacity allows to specify utxo cache size (otherwise set to DEFAULT_CAPACITY)
get_random_utxo returns a Result<Option<OutPoint>>, as some seeds don't produce any utxo.

danielabrozzoni committed:
Fix error handling

afilini commented on src/demo.rs:
```diff
@@ -22,45 +21,98 @@ where
+        Ok(unspent_utxos
+            .into_iter()
+            .filter(|x| x.tx_hash == txout.txid)
+            .count()
```
You can probably use `any(...)` as a shortcut

afilini commented on src/demo.rs:
```diff
@@ -22,45 +21,98 @@ where
+                // We hit a coinbase!
+                if txid == &coinbase_txid {
+                    break;
+                }
```
There's a method called `is_coin_base()` on `bitcoin::Transaction` for that

afilini commented on src/demo.rs:
```diff
@@ -22,45 +21,98 @@ where
+                scripts.extend(
+                    tx.output
+                        .iter()
+                        .map(|x| x.script_pubkey.clone())
```
Since you never use `tx` afterwards, you could use `into_iter()` and then avoid cloning the scripts.

danielabrozzoni replied:
Nope: since I'm using `tx.is_coin_base` (instead of `txid == coinbase`, as you suggested), tx is used at the next iteration of the loop, thus it can't be partially moved using `into_iter`

afilini commented on src/demo.rs:
```diff
@@ -22,45 +21,98 @@ where
+                    tx.output
+                        .iter()
+                        .map(|x| x.script_pubkey.clone())
+                        .filter(|x| x.is_p2pkh())
```
Should be `is_p2wpkh()`, that's what we want

danielabrozzoni committed:
Cleanup

danielabrozzoni committed:
Privacy improvements

Randomly navigatin utxos from the seed, making get_random_utxo calls non
deterministic from the seed

danielabrozzoni committed:
Add u64 seed in generate_utxo

danielabrozzoni committed:
Cargo fmt

afilini commented on src/demo.rs:
```diff
@@ -1,59 +1,138 @@
+        seed: u64,
+    ) -> Result<Option<OutPoint>, Self::Error> {
+        let mut rng: StdRng = SeedableRng::seed_from_u64(seed);
+        if self.utxo_set.borrow().len() == 0 {
```
`is_empty()`

afilini commented on src/demo.rs:
```diff
@@ -1,59 +1,138 @@
+                        .collect::<Vec<_>>(),
+                );
+
+                for script in &scripts {
```
it looks like you never empty `scripts`, so i think you keep asking for the same scripts every time even if you have already checked them. You should maybe use one global set and a local one, make the difference and once you are done add all the elements from the local set to the global one.

afilini commented on src/demo.rs:
```diff
@@ -1,59 +1,138 @@
+                    let unspent: Vec<_> = self
+                        .electrum_client
+                        .borrow_mut()
+                        .script_list_unspent(&script)?
```
You should try to use the batching functions to make this more efficient

afilini commented on src/demo.rs:
```diff
@@ -1,59 +1,138 @@
+                            vout: x.tx_pos as u32,
+                        })
+                        .collect();
+                    self.utxo_set.borrow_mut().extend(unspent);
```
I'm not sure it's a good idea to take all the unspent outputs of a tx: imagine the case where my utxo is not the only unspent output in its previous transaction. Just by the fact that the other outputs in the same tx weren't selected in the random set I can find it with pretty high accuracy.

afilini commented on src/demo.rs:
```diff
@@ -1,59 +1,138 @@
+            let mut tx = self.get_tx(&txid)?.clone();
+            let mut scripts = HashSet::new();
+
+            for _ in 0..rng.gen_range(1, 3) {
```
I would do a few more than 3, maybe in the 10-20 range, idk

danielabrozzoni replied:
IMHO we should consider how many steps you have to do, on average, before hitting a coinbase. Any idea how to calculate it?
My reason for putting (1, 3) was: I have less chances to hit a coinbase; but I understand this could leak privacy.

afilini commented on src/demo.rs:
```diff
@@ -1,59 +1,138 @@
+            for _ in 0..rng.gen_range(1, 3) {
+                // We hit a coinbase!
+                if tx.is_coin_base() {
+                    break;
```
maybe you should return here? I'm thinking about the case where you almost immediately hit a coinbase and so you didn't make that many steps back. Better to fail and retry with a different seed

afilini commented on src/demo.rs:
```diff
@@ -1,59 +1,138 @@
+                    break;
+                }
+
+                txid = &tx.input[rng.gen_range(0, &tx.input.len())]
```
I think you can use [`choose`](https://docs.rs/rand/0.7.3/rand/seq/trait.SliceRandom.html#tymethod.choose) to pick a random element from a slice

afilini commented on src/demo.rs:
```diff
@@ -1,59 +1,138 @@
+                    break;
+                }
+
+                txid = &tx.input[0].previous_output.txid;
```
Is there a reason why 0 is hardcoded here?

danielabrozzoni committed:
WIP new algorithm
